### PR TITLE
WIP: Restrict Prometheus instance namespaces

### PR DESCRIPTION
**Pull request by redbaron**
Fixes https://github.com/coreos/prometheus-operator/issues/2186

WIP for comments and e2e tests

### CODE REVIEW

File: cmd/operator/main.go [hunk 1]
```diff
@@ -125,6 +126,7 @@ func init() {
 	flagset.StringVar(&cfg.PrometheusDefaultBaseImage, "prometheus-default-base-image", "quay.io/prometheus/prometheus", "Prometheus default base image")
 	flagset.StringVar(&cfg.ThanosDefaultBaseImage, "thanos-default-base-image", "improbable/thanos", "Thanos default base image")
 	flagset.Var(ns, "namespaces", "Namespaces to scope the interaction of the Prometheus Operator and the apiserver.")
+	flagset.Var(promNS, "prometheus-instance-namespaces", "Namespaces where Prometheus CRDs and corresponding Secrets,Configmaps and StatefulSets are watched/created.")
```
Comment (brancz) [inline cmd/operator/main.go]: I think I'd prefer it to be the other way around.

Comment (redbaron) [reply cmd/operator/main.go]: see my comments in https://github.com/coreos/prometheus-operator/issues/2186

**Comment by redbaron** [comment]
@brancz , it should be all green, please have a look

File: test/e2e/prometheus_test.go [hunk 2]
```diff
@@ -1553,6 +1553,54 @@ func testOperatorNSScope(t *testing.T) {
...
+		p.Spec.EvaluationInterval = "1s"
+		p, err = framework.CreatePrometheusAndWaitUntilReady(prometheusNS, p)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		pMissed := framework.MakeBasicPrometheus(ruleNS, name, name, 1)
+		pMissed.Spec.EvaluationInterval = "1s"
+		pMissed, err = framework.CreatePrometheusAndWaitUntilReady(ruleNS, pMissed)
+		if err == nil {
```
Comment (redbaron) [inline test/e2e/prometheus_test.go]: I am not sure how to test that it ignores Prometheus CR in a given namespace. I didn't come up with anything better than checking that error (timeout) happens indeed, but it makes test run longer :(

**Comment by redbaron** [comment]
Depends on https://github.com/coreos/prometheus-operator/pull/1997  to support BasicAuth secrets for endpoints

**Comment by redbaron** [comment]
WIP until single-item watches added for Secrets in SMon endpoints. Otherwise their updates are not seen right now

**Comment by brancz** [comment]
@redbaron with https://github.com/coreos/prometheus-operator/pull/2202 being merged, do you want to pick this up?

**Comment by s-urbaniak** [comment]
@redbaron just fyi: i am working actively now on this feature :-) feel free to chime in if you are still planning to work on this too.

File: cmd/operator/main.go (hunk 1 above)
Comment (squat) [reply cmd/operator/main.go]: Nit: `Secrets,Configmaps` -> `Secrets, ConfigMaps`

File: pkg/prometheus/operator.go [hunk 3]
```diff
@@ -144,6 +148,7 @@ type Config struct {
 	PrometheusDefaultBaseImage    string
 	ThanosDefaultBaseImage        string
 	Namespaces                    []string
+	PromInstanceNamespaces        []string //  limits where Prometheus CRDs,their secrets and configmaps are watched for
```
Comment (squat) [inline pkg/prometheus/operator.go]: Nit: add space after comma

**Comment by s-urbaniak** [comment]
This is effectively replaced by https://github.com/coreos/prometheus-operator/pull/2783, minus the embedded secret informer framework.

**Comment by redbaron** [comment]
@s-urbaniak, I might have missed some details, but although it is similar it is not like for like replacement. #2783 requires clusterrolebinding to be able to watch all resources in all namespaces, it then choses to ignore some, but service account permissions are still there. With this PR operator can start and work without cluster-wide access

**Comment by s-urbaniak** [comment]
@redbaron indeed, you are absolutely right. #2783 is not a drop-in replacement for this PR, just a subset filtering prometheus/alertmanager custom resources.

**Pull request closed without merging**